sklearn/linear_model/ridge.py

```diff
@@ -141,7 +145,16 @@ def _solve_cholesky_kernel(K, y, alpha, sample_weight=None):
         # Only one penalty, we can solve multi-target problems in one time.
         K.flat[::n_samples + 1] += alpha[0]
 
-        dual_coef = linalg.solve(K, y, sym_pos=True, overwrite_a=True)
+        try:
+            # Note: we must use overwrite_a=False in order to be able to
+            #       use the fall-back solution below in case a LinAlgError
+            #       is raised
+            dual_coef = linalg.solve(K, y, sym_pos=True,
+                                     overwrite_a=False)
+        except np.linalg.LinAlgError:
+            warnings.warn("Singular matrix in solving dual problem. Using "
+                          "least-squares solution instead.")
+            dual_coef = linalg.lstsq(K, y)[0]
```

Comment:
beautiful! thanks for the effort!!

On Wednesday, January 14, 2015, Jan Hendrik Metzen <notifications@github.com>
wrote:

> In sklearn/linear_model/ridge.py
> <https://github.com/scikit-learn/scikit-learn/pull/3942#discussion-diff-22955171>
> :
>
> > @@ -141,7 +145,16 @@ def _solve_cholesky_kernel(K, y, alpha, sample_weight=None):
> >          # Only one penalty, we can solve multi-target problems in one time.
> >          K.flat[::n_samples + 1] += alpha[0]
> >
> > -        dual_coef = linalg.solve(K, y, sym_pos=True, overwrite_a=True)
> > +        try:
> > +            # Note: we must use overwrite_a=False in order to be able to
> > +            #       use the fall-back solution below in case a LinAlgError
> > +            #       is raised
> > +            dual_coef = linalg.solve(K, y, sym_pos=True,
> > +                                     overwrite_a=False)
> > +        except np.linalg.LinAlgError:
> > +            warnings.warn("Singular matrix in solving dual problem. Using "
> > +                          "least-squares solution instead.")
> > +            dual_coef = linalg.lstsq(K, y)[0]
>
> Out of curiosity, I did a comparison ot the execution time of
> "scipy.linalg.solve(K, y, sym_pos=True, overwrite_a=False)" and
> "scipy.linalg.lstsq(K, y)[0]" for different sizes of K:
> [image: figure_1]
> <https://cloud.githubusercontent.com/assets/1116263/5744184/a78f32b4-9c1e-11e4-9dcc-3151d3d9a332.png>
> solve is thus between one and two orders of magnitude faster.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/3942/files#r22955171>.
>